openmc/lattice.py

```diff
@@ -0,0 +1,869 @@
+            return False
+        elif self.id != other.id:
+            return False
+        elif self.name != other.name:
+            return False
+        elif self.pitch != other.pitch:
+            return False
+        elif self.outer != other.outer:
+            return False
+        elif self.universes != other.universes:
+            return False
+        else:
+            return True
+
+    def __ne__(self, other):
+        return not self == other
+
+    @property
+    def id(self):
+        return self._id
+
+    @property
+    def name(self):
+        return self._name
+
+    @property
+    def pitch(self):
+        return self._pitch
+
+    @property
+    def outer(self):
+        return self._outer
+
+    @property
+    def universes(self):
+        return self._universes
+
+    @id.setter
+    def id(self, lattice_id):
+        if lattice_id is None:
+            global AUTO_UNIVERSE_ID
+            self._id = AUTO_UNIVERSE_ID
+            AUTO_UNIVERSE_ID += 1
+        else:
+            cv.check_type('lattice ID', lattice_id, Integral)
+            cv.check_greater_than('lattice ID', lattice_id, 0, equality=True)
+            self._id = lattice_id
+
+    @name.setter
+    def name(self, name):
+        if name is not None:
+            cv.check_type('lattice name', name, basestring)
+            self._name = name
+        else:
+            self._name = ''
+
+    @outer.setter
+    def outer(self, outer):
+        cv.check_type('outer universe', outer, Universe)
+        self._outer = outer
+
+    @universes.setter
+    def universes(self, universes):
+        cv.check_iterable_type('lattice universes', universes, Universe,
+                               min_depth=2, max_depth=3)
+        self._universes = np.asarray(universes)
+
+    def get_unique_universes(self):
+        """Determine all unique universes in the lattice
+
+        Returns
+        -------
+        universes : collections.OrderedDict
+            Dictionary whose keys are universe IDs and values are Universe
+            instances
+
+        """
+
+        univs = OrderedDict()
+        for k in range(len(self._universes)):
+            for j in range(len(self._universes[k])):
+                if isinstance(self._universes[k][j], Universe):
+                    u = self._universes[k][j]
+                    univs[u._id] = u
+                else:
+                    for i in range(len(self._universes[k][j])):
+                        u = self._universes[k][j][i]
+                        assert isinstance(u, Universe)
+                        univs[u._id] = u
+
+        if self.outer is not None:
+            univs[self.outer._id] = self.outer
+
+        return univs
+
+    def get_all_nuclides(self):
+        """Return all nuclides contained in the lattice
+
+        Returns
+        -------
+        nuclides : collections.OrderedDict
+            Dictionary whose keys are nuclide names and values are 2-tuples of
+            (nuclide, density)
```

Comment:
This one I'm going to leave as is -- if I'm referring to type I'd have to say `(openmc.Nuclide, float)` which seems weird.